coalib/parsing/CliParsing.py

```diff
@@ -78,6 +78,65 @@ def parse_cli(arg_list=None,
+                        arg_parser,
+                        args,
+                        key_value_delimiters=('=', ':'),
+                        comment_seperators=(),
+                        key_delimiters=(',',),
+                        section_override_delimiters=('.',),
+                        key_value_append_delimiters=('+=',)):
+    """
+    Parses the custom settings from CLI arguments and returns them.
+
+    :param arg_list:                    The CLI argument list.
+    :param arg_parser:                  Instance of ArgParser that is used to
+                                        parse none-setting arguments.
+    :param args:                        Alternative pre-parsed CLI arguments.
+    :param key_value_delimiters:        Delimiters to separate key and value
+                                        in setting arguments where settings are
+                                        being defined.
+    :param comment_seperators:          Allowed prefixes for comments.
+    :param key_delimiters:              Delimiter to separate multiple keys of
+                                        a setting argument.
+    :param section_override_delimiters: The delimiter to delimit the section
+                                        from the key name (e.g. the '.' in
+                                        sect.key = value).
+    :param key_value_append_delimiters: Delimiters to separate key and value
+                                        in setting arguments where settings are
+                                        being appended.
+    :return:                            A list of dicts with a setting key to
+                                        denote the name of a custom setting,
+                                        and a value key to denote the value
+                                        of that custom setting.
+    """
+    assert not (arg_list and args), (
+        'Either call get_custom_settings() with an arg_list of CLI arguments '
+        'or with pre-parsed args, but not with both.')
+
+    if args is None:
+        arg_parser = default_arg_parser() if arg_parser is None else arg_parser
+        args = arg_parser.parse_args(arg_list)
+
+    line_parser = LineParser(key_value_delimiters,
+                             comment_seperators,
+                             key_delimiters,
+                             {},
+                             section_override_delimiters,
+                             key_value_append_delimiters)
+
+    parsed_settings = []
+    for arg_key, arg_value in sorted(vars(args).items()):
+        if arg_key == 'settings' and arg_value is not None:
+            for setting_definition in arg_value:
+                (_, key_tuples, value, _, _) = line_parser._parse(
+                    setting_definition)
+                for key_tuple in key_tuples:
+                    parsed_settings.append(
+                        {'setting': key_tuple[1], 'value': value})
```

Comment:
hmmm... much of this code seems duplicated from the function above and function below, the function above does half of the job, and gives it to the function below, which calls `append_to_sections()` upon it, while this function just seems to be returning those settings. Can this be done without code duplication with additional params and changes to the functions already present?

Reply:
or if you really need these setting values in `ConfigurationGathering.py` maybe you can obtain those from section values that the function `parse_custom_settings()` is appending to.

Reply:
With how the methods are currently written, I haven't been able to take a section and determine which of the settings associated with it are custom. `parse_cli()` determines which settings are custom settings from the command line and then calls `parse_custom_settings()`, which appends a custom setting to the section. However, once it has been appended to the section, there is no indication of if it is a custom setting or not.

One idea I had was to modify one of the existing methods (probably `parse_cli()`) and have it return an extra parameter, e.g., modify `parse_cli()` so that it returns the sections and a list of custom settings. However, this would require modifying the signatures and outputs of several other methods as well.

I think it makes the most sense to modify the Setting object so that it has an "is_custom" attribute. Thoughts? @jayvdb